ExtCore/Control.Cps.Result.fs

```diff
@@ -0,0 +1,265 @@
+        : ResultContFunc<'T, 'Error, 'K> =
+        fun cont ->
+            cont (Ok value)
+
+    // M<'T> -> M<'T>
+    member inline __.ReturnFrom func
+        : ResultContFunc<'T, 'Error, 'K> =
+        func
+
+    // unit -> M<'T>
+    member inline __.Zero ()
+        : ResultContFunc<unit, 'Error, 'K> =
+        fun cont ->
+            cont (Ok ())
+
+    // (unit -> M<'T>) -> M<'T>
+    member __.Delay f
+        : ResultContFunc<'T, 'Error, 'K> =
+        f ()
+
+    // M<'T> * ('T -> M<'U>) -> M<'U>
+    member inline __.Bind (m : ResultContFunc<_,_,_>, k : 'T -> ResultContFunc<_,_,_>)
+        : ResultContFunc<'U, 'Error, 'K> =
+        fun cont ->
+            m <| fun result ->
+            match result with
+            | Error error ->
+                Error error
+                |> cont
+            | Ok value ->
+                k value cont    
+
+    // M<'T> -> M<'T> -> M<'T>
+    // or
+    // M<unit> -> M<'T> -> M<'T>
+    member inline __.Combine (r1 : ResultContFunc<_,_,_>, r2 : ResultContFunc<_,_,_>)
+        : ResultContFunc<'T, 'Error, 'K> =
+        fun cont ->
+            r1 <| fun result ->
+            match result with
+            | Error error ->
+                Error error
+                |> cont
+            | Ok () ->
+                r2 cont
+
+    // M<'T> * (exn -> M<'T>) -> M<'T>
+    member inline __.TryWith (body : ResultContFunc<_,_,_>, handler : exn -> ResultContFunc<_,_,_>)
+        : ResultContFunc<'T, 'Error, 'K> =
+        fun cont ->
+            try body cont
+            with ex ->
+                handler ex cont
+
+    // M<'T> -> M<'T> -> M<'T>
+    member inline __.TryFinally (body : ResultContFunc<_,_,_>, handler)
+        : ResultContFunc<'T, 'Error, 'K> =
+        fun cont ->
+            try body cont
+            finally
+                handler ()
+
+    // 'T * ('T -> M<'U>) -> M<'U> when 'U :> IDisposable
+    member this.Using (resource : ('T :> System.IDisposable), body : 'T -> ResultContFunc<_,_,_>)
+        : ResultContFunc<'U, 'Error, 'K> =
+        this.TryFinally (body resource, (fun () ->
+            if not <| isNull (box resource) then
+                resource.Dispose ()))
+
+    // (unit -> bool) * M<'T> -> M<'T>
+    member this.While (guard, body : ResultContFunc<_,_,_>)
+        : ResultContFunc<unit, 'Error, 'K> =
+        if guard () then
+            this.Bind (body, (fun () -> this.While (guard, body)))
+        else
+            this.Zero ()
+
+    // seq<'T> * ('T -> M<'U>) -> M<'U>
+    // or
+    // seq<'T> * ('T -> M<'U>) -> seq<M<'U>>
+    member this.For (sequence : seq<_>, body : 'T -> ResultContFunc<_,_,_>)
+        : ResultContFunc<unit, 'Error, 'K> =
+        this.Using (sequence.GetEnumerator (), fun enum ->
+            this.While (
+                enum.MoveNext,
+                this.Delay (fun () ->
+                    body enum.Current)))
+
+
+/// <summary>
+/// </summary>
+[<Sealed>]
+type ProtectedResultStateContBuilder () =
```

Comment:
I don't like the name. The "choice" version is named just "ProtectedState..." where "protected" means Either-style error handling. Here we have a kind of duplication: "protected" and "result" both point to the same aspect of the builder - errors handling.

Reply:
Indeed, should we move the "choice" versions to "compatibility" namespace, so that the result versions can have the same name ?